tests/api/fuzzer.py

```diff
@@ -0,0 +1,319 @@
+import argparse
+import json
+import logging
+import random
+import re
+import requests
+import sys
+import posixpath
+import urllib.parse
+
+
+#######################################################################################################################
+### Utilities
+
+def some(s):
+    return random.choice(sorted(s))
+
+def pretty_json(tree, depth=0, max_depth=None):
+    if max_depth is not None  and  depth>=max_depth:
+        return
+    indent = "  "*depth
+    for k,v in tree.items():
+        if isinstance(v, int) or isinstance(v, str):
+            print(f"{indent}{k}: {v}")
+        elif isinstance(v, dict):
+            print(f"{indent}{k}: ->")
+            pretty_json(v, depth+1)
+        else:
+            print(f"{indent}{k}: {type(v)}")
+
+#######################################################################################################################
+### Data-model and processing
+
+class Param(object):
+    def __init__(self, name, location, kind):
+        self.location = location
+        self.kind     = kind
+        self.name     = name
+        self.values   = set()
+
+    def dump(self):
+        print(f"{self.name} in {self.location} is {self.kind} : {{{self.values}}}")
+
+def does_response_fit_schema(schema_path, schema, resp):
+    '''The schema_path argument tells us where we are (globally) in the schema. The schema argument is the
+       sub-tree within the schema json that we are validating against. The resp is the json subtree from the
+       target host's response.
+
+       The basic idea is this: swagger defines a model of valid json trees. In this sense it is a formal
+       language and we can validate a given server response by checking if the language accepts a particular
+       server response. This is basically a parser, but instead of strings we are operating on languages
+       of trees.
+
+       This could probably be extended to arbitrary swagger definitions - but the amount of work increases
+       rapidly as we attempt to cover the full semantics of languages of trees defined in swagger. Instead
+       we have some special cases that describe the parts of the semantics that we've used to describe the
+       netdata API.
+    '''
+    if "type" in schema  and  schema["type"]=="object":
```

Comment:
I would split the `type` key presence check from the value check so as not to repeat it every time which would result in another nested if.

Reply:
Yes, there is lots of duplication in this method. I'm still trying to work out what the set of common operations on the schema looks like and then I will clean it up.